### ISSUE TITLE
Is there any plan to implement an option to deliver data out of order?

### ISSUE THREAD

**Issue by ryskiwt** [quic-go/quic-go#1518]
Dear maintainers,

QUIC draft says that implemantations can choose to offer the ability to deliver data out of order.

https://tools.ietf.org/html/draft-ietf-quic-transport-14#section-9

>    Stream offsets allow for the octets on a stream to be placed in
>    order.  An endpoint MUST be capable of delivering data received on a
>    stream in order.  Implementations MAY choose to offer the ability to
>    deliver data out of order.  There is no means of ensuring ordering
>    between octets on different streams.

Is there any plan to implement an option to deliver data out of order?
I think this option extends applicability of QUIC to delay-sensitive applications such as media data transfer. 

**Just for your information:**
I forked and tried adding very simple implementation. (https://github.com/ryskiwt/quic-go/compare/759051dae9ebff2b5a2a0deda818834991829c73...ryskiwt:feature/disable_sort)

I would be happy if you will consider introducing such option.

Thanks,

Ryosuke

**Comment by marten-seemann** [comment]
The Stream interface implements io.Reader. io.Reader implies in-order delivery. For out-of-order delivery, we’d need to add another method to the Stream (or another callback).
I’m not opposed to doing this, if someone comes up with a specific use case for this.

**Comment by jared2501** [comment]
I would love the ability to have an out-of-order stream! Additionally, it
would be good to impleme the non-guarenteed delivery part of the spec too.
The use case I have is streaming video on a quic stream. Over the same
connection I'd like to have an ordered, reliable stream of metadata at a
high priority, and an unordered non-reliabe video stream. Happy to answer
more questions about the use case!

**Comment by marten-seemann** [comment]
Hi @jared2501, which non-guaranteed delivery part of the spec do you mean? Stream data is always delivered reliably, unless the stream is reset.
Would an API that looks something like 
```go
Stream.OnDataReceived() (uint64 /* offset */, []byte, error)
```
work for your use case?

**Comment by jared2501** [comment]
This spec: https://tools.ietf.org/html/draft-tiesel-quic-unreliable-http-00

Although I haven't been tracking the spec closely. Did this draft make it
in?

And that interface looks reasonable to me.

**Comment by marten-seemann** [comment]
No, that's not part of the QUIC spec. All QUIC streams are reliable. Partial reliability was postponed to QUIC v2.

**Comment by ryskiwt** [comment]
Hi @marten-seemann,

I'm now developing a message broker to stream high-frequency sensor data using QUIC stream.
(My use case is somewhat similar to the use case of @jared2501)

In networks with delays, the feature of QUIC's in-order delivery sometimes causes a blocking of bytes. (Bytes contained in a suddenly delayed packet may block subsequent bytes)

Since I have built my own framing mechanism in a layer higher than QUIC,
it does not matter that the bytes order is not guaranteed.
I just want to get received bytes as soon as it's delivered by the QUIC stream.

So, the interface you suggested is what I'm wanted.

Thanks

**Comment by marten-seemann** [comment]
There's a new proposal to add a MESSAGE frame to QUIC in https://mailarchive.ietf.org/arch/msg/quic/p0SSsnlUi_m6xNvTmBEJ41UPYA0. It sounds like this would be the better solution for your use case, since you don't care about reliable delivery, right?

With the current implementation, why don't you just use a new stream for every measurement? That would eliminate all head-of-line-blocking.

**Comment by ryskiwt** [comment]
I think that the proposal you mentioned matches my use case.
I will follow the updates. Thank you.


> why don't you just use a new stream for every measurement?

What you mean is to treat a stream something like a message, isn't it?
The usage may work for my use case, but there is one thing I'm worring about.

The data I'm now trying to transmit is somewhat frequent. (transmit messages every 1 ms to 10 ms) 
Can this library withstand the usage that opens and closes 1000 streams per second?
I'm worring about the costs to open and close streams.

**Comment by ydnar** [comment]
This is a more recently-updated proposal to add unreliable datagrams to QUIC: https://github.com/tfpauly/draft-pauly-quic-datagram

**Comment by marten-seemann** [comment]
Out-of-order delivery has nothing to do with the datagram draft. Datagrams are about unreliable delivery.

**Comment by go101** [comment]
I am curious that has Datagrams been supported in quic-go?

**Comment by marten-seemann** [comment]
Closing this issue, since we have support for DATAGRAMs now (for quite a while already, actually).

While this is NOT out-of-order delivery of stream data, it sounds like most use cases for out-of-order delivery can be solved by switching to datagrams.
